libs/vkd3d/command.c

```diff
@@ -6329,6 +6334,45 @@ static void STDMETHODCALLTYPE d3d12_command_list_CopyTextureRegion(d3d12_command
         return;
 
     d3d12_command_list_ensure_transfer_batch(list, copy_info.batch_type);
+    alias = false;
+    for (i = 0; !alias && i < list->transfer_batch.batch_len; i++)
+    {
+        const struct vkd3d_image_copy_info *other_info = &list->transfer_batch.batch[i];
+        const VkImageSubresourceLayers *subres, *other_subres;
+        const struct d3d12_resource *res, *other_res;
+
+        switch (copy_info.batch_type)
+        {
+            case VKD3D_BATCH_TYPE_COPY_BUFFER_TO_IMAGE:
+                /* Test for destination aliasing as D3D12 requires serialization on overlapping copies (WAW hazard). */
+                subres = &copy_info.copy.buffer_image.imageSubresource;
+                other_subres = &other_info->copy.buffer_image.imageSubresource;
+                assert(subres->layerCount == 1 && other_subres->layerCount == 1);
+                alias = copy_info.dst.pResource == other_info->dst.pResource &&
+                         subres->aspectMask == other_subres->aspectMask &&
+                         subres->baseArrayLayer == other_subres->baseArrayLayer &&
+                         subres->mipLevel == other_subres->mipLevel;
+                break;
+            case VKD3D_BATCH_TYPE_COPY_IMAGE_TO_BUFFER:
+                /* Test for destination aliasing as D3D12 requires serialization on overlapping copies (WAW hazard). */
+                /* TODO: Do more granular alias testing or merge this with d3d12_command_list_mark_copy_buffer_write. */
+                res = impl_from_ID3D12Resource(copy_info.dst.pResource);
+                other_res = impl_from_ID3D12Resource(other_info->dst.pResource);
+                /* If either resource is sparse, consider it to alias with anything. */
+                alias = copy_info.dst.pResource == other_info->dst.pResource ||
+                        (res->flags & VKD3D_RESOURCE_RESERVED) || (other_res->flags & VKD3D_RESOURCE_RESERVED);
+                break;
+            case VKD3D_BATCH_TYPE_COPY_IMAGE:
+                /* TODO: Check for alias once we start batching barriers for image-image copies too */
+                break;
+            default:
+                assert(false);
+                return;
+        }
+    }
+    if (alias)
+        d3d12_command_list_end_transfer_batch(list);
```

Comment:
If we end the transfer batch here, wouldn't batch type be reset to NONE, and thus the copy_info we append here will be dropped?

Reply:
Ouch, it doesn’t get dropped because we don’t reset length for NONE, which is why I didn’t catch it in testing, but it will cause legitimate issue when the batch length is 1. I’ll move the ensure() down to prevent that.